### PR TITLE
Fix #4477: Build fails after building specific files

### LINKED ISSUE #4477: Build fails after building specific files
### Problem
If you pass `[FILENAMES, ...]` to `sphinx-build`, a subsequent call _without_ this argument will fail.

#### Procedure to reproduce the problem
```
$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/
$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/ doc/ext/math.rst
$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/
```

#### Error logs / results
```
$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/                                      
Running Sphinx v1.8.0+/9a3f401c4                                                                         
making output directory...                                                                                                                                                                                         
building [mo]: targets for 0 po files that are out of date                                               
building [html]: targets for 70 source files that are out of date                                        
updating environment: 70 added, 0 changed, 0 removed
reading sources... [100%] websupport                
looking for now-outdated files... none found        
pickling environment... done                        
checking consistency... done                        
preparing documents... done                         
writing output... [100%] websupport                 
generating indices... genindex py-modindex                                                               
highlighting module code... [100%] sphinxcontrib.websupport.storage                                      
writing additional pages... index search opensearch 
copying images... [100%] themes/alabaster.png                                                            
copying downloadable files... [100%] ext/example_google.py                                               
copying static files... done                        
copying extra files... done                         
dumping search index in English (code: en) ... done 
dumping object inventory... done                    
build succeeded.                                    
                                                    
The HTML pages are in build/sphinx/html.

$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/ doc/ext/math.rst
Running Sphinx v1.8.0+/9a3f401c4
building [mo]: targets for 0 po files that are specified
building [html]: 1 source files given on command line
updating environment: 70 added, 0 changed, 0 removed
reading sources... [100%] websupport
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] ext/math
generating indices... genindex py-modindex
highlighting module code... [100%] sphinxcontrib.websupport.storage
writing additional pages... index search opensearch
copying downloadable files... [100%] ext/example_google.py
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.

The HTML pages are in build/sphinx/html.

$ sphinx-build -E doc/ -d build/sphinx/doctrees/ build/sphinx/html/
Running Sphinx v1.8.0+/9a3f401c4
building [mo]: targets for 0 po files that are out of date
WARNING: unsupported build info format in u'/home/sfinucan/Development/sphinx/sphinx/build/sphinx/html/.buildinfo', building all
building [html]: targets for 70 source files that are out of date
updating environment: 70 added, 0 changed, 0 removed
reading sources... [100%] websupport
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] websupport
generating indices... genindex py-modindex
highlighting module code... [100%] sphinxcontrib.websupport.storage
writing additional pages... index search opensearch
copying images... [100%] themes/alabaster.png
copying downloadable files... [100%] ext/example_google.py
copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded, 1 warnings.

The HTML pages are in build/sphinx/html.
```

The issue is this line in the log of the third call to `sphinx-build`.

```
WARNING: unsupported build info format in u'/home/sfinucan/Development/sphinx/sphinx/build/sphinx/html/.buildinfo', building all
```

#### Expected results
There should be no warnings and all files should be rebuilt.

### Reproducible project / your project
- Sphinx docs

### Environment info
- OS: Fedora 27
- Python version: Python 2.7
- Sphinx version: master (9a3f401c469ec6c16be32ce1bda11d266ba88d76)

### PR DESCRIPTION
**Pull request by tk0miya**
refs: #4477 and #4478

### CODE REVIEW

File: sphinx/builders/html.py [hunk 1]
```diff
@@ -147,6 +149,56 @@ def __new__(cls, filename, title, rel):
         return self
 
 
+class BuildInfo(object):
+    """buildinfo file manipulator.
+
+    HTMLBuilder and its family are storing its own envdata to ``.buildinfo``.
```
Comment (jfbu) [inline sphinx/builders/html.py]: s/its own/their own/

File: sphinx/builders/html.py [hunk 2]
```diff
@@ -147,6 +149,56 @@ def __new__(cls, filename, title, rel):
...
+            assert lines[0].rstrip() == '# Sphinx build info version 1'
+            assert lines[2].startswith('config: ')
+            assert lines[3].startswith('tags: ')
+
+            build_info = BuildInfo()
+            build_info.config_hash = lines[2].split()[1].strip()
+            build_info.tags_hash = lines[3].split()[1].strip()
+            return build_info
+        except Exception as exc:
+            raise ValueError('build info file has broken: %r' % exc)
```
Comment (jfbu) [inline sphinx/builders/html.py]: s/has/is/

**Review by jfbu — APPROVED** [review]
looks good to me but I don't yet know enough of Sphinx internals; I had no idea that html like builders did their own buildinfo manipulation.

**Comment by tk0miya** [comment]
just fixed typos

**Pull request merged**